pkg/counter/cache.go

```diff
@@ -175,24 +204,77 @@ func indexOfMember(members []toolchainv1alpha1.Member, name string) int {
 	return -1
 }
 
-func shouldLoadCurrentResources(toolchainStatus *toolchainv1alpha1.ToolchainStatus) bool {
-	return (toolchainStatus.Status.HostOperator == nil ||
-		toolchainStatus.Status.HostOperator.MasterUserRecordCount == 0) &&
-		!cachedCounts.initialized
+func initialize(cl client.Client, toolchainStatus *toolchainv1alpha1.ToolchainStatus) error {
+	// skip if cached counters are already initialized
+	if cachedCounts.initialized {
+		return nil
+	}
+
+	// initialize the cached counters from the UserSignup and MasterUserRecord resources.
+	if toolchainStatus.Status.HostOperator == nil ||
+		toolchainStatus.Status.HostOperator.MasterUserRecordCount == 0 {
+		return initializeFromResources(cl, toolchainStatus.Namespace)
+	}
+	// otherwise, initialize the cached counters from the ToolchainStatud resource.
+	return initializeFromToolchainStatus(cl, toolchainStatus)
 }
 
-func loadCurrentResources(cl client.Client, namespace string) error {
+// initialize the cached counters from the UserSignup and MasterUserRecord resources.
+// this func lists all UserSignup and MasterUserRecord resources
+func initializeFromResources(cl client.Client, namespace string) error {
+	log.Info("initializing counters from resources")
+	usersignups := &toolchainv1alpha1.UserSignupList{}
+	if err := cl.List(context.TODO(), usersignups, client.InNamespace(namespace)); err != nil {
+		return err
+	}
 	murs := &toolchainv1alpha1.MasterUserRecordList{}
 	if err := cl.List(context.TODO(), murs, client.InNamespace(namespace)); err != nil {
 		return err
 	}
 	reset()
+	for _, usersignup := range usersignups.Items {
```

Comment:
Let's make sure there is no race condition when we start migrating the existing UserSignups (sets annotation and increments the toolchain status) and when we initialize the cache from resources.

Reply:
done in 594ea21, using a public lock obtained by the `counter.initialize()` func and the `UserSignupController` during its migration.

Reply:
reverted change after discussing about it with @MatousJobanek on https://coreos.slack.com/archives/CHK0J6HT6/p1618836017462800?thread_ts=1618828935.454500&cid=CHK0J6HT6